lxd/images.go

```diff
@@ -4010,43 +4043,82 @@ func imageExport(d *Daemon, r *http.Request) response.Response {
 		return response.SmartError(err)
 	}
 
-	// Get the image (expand the fingerprint).
+	isDevLXDQuery := r.RemoteAddr == "@devlxd"
+	secret := r.FormValue("secret")
+	trusted := auth.IsTrusted(r.Context())
+
+	// Unauthenticated clients that do not provide a secret may only view public images.
+	// Devlxd queries can be for private images but only if they are cached.
```

Comment:
Why are only private cached images allowed? I feel like this comment could do with expanding.

Reply:
I've updated the comments here but I'm still not sure I've made it super clear to be honest.

Devlxd queries can be for public images because the image is accessible to anyone. Devlxd queries can be for private images only when the image is cached from a public remote. E.g. if the host has `ubuntu:22.04` downloaded, we cache it but don't make it public.

Reply:
Thanks, that makes sense.

We don't want to trigger a download of an image on the host.

BTW are we restricting which project images are accessible by /dev/lxd requests to only those of the instance's effective image project?

Reply:
Currently are devlxd requests are for the default project.

I think this warrants a separate issue though. As I understand it, all cached images are stored in the default project. If I launch an `ubuntu:22.04` container in project `test` say, and then create another `ubuntu:22.04` inside it, then restricting the `devlxd` request to only project `test` would miss the cached image and download it via the ubuntu remote instead.

Reply:
Another thing that doesn't make sense to me is how to actually make an image public on a host that is accessible to the nested LXD. The nested LXD doesn't have the host added as a remote, so it can't resolve the alias of the public image. As far as I can tell, this feature only works for cached images.

Reply:
OK thanks, yes please can you create a separate issue. 

I'm not sure that "all cached images are stored in the default project" but if /dev/lxd only servers from the default project currently then we dont need to change it in this PR.

Reply:
> Another thing that doesn't make sense to me is how to actually make an image public on a host that is accessible to the nested LXD. The nested LXD doesn't have the host added as a remote, so it can't resolve the alias of the public image. As far as I can tell, this feature only works for cached images.

Does `lxc import server:fingerprint` work?

Reply:
> OK thanks, yes please can you create a separate issue.
> 
> I'm not sure that "all cached images are stored in the default project" but if /dev/lxd only servers from the default project currently then we dont need to change it in this PR.

https://github.com/canonical/lxd/issues/13768

> > Another thing that doesn't make sense to me is how to actually make an image public on a host that is accessible to the nested LXD. The nested LXD doesn't have the host added as a remote, so it can't resolve the alias of the public image. As far as I can tell, this feature only works for cached images.
> 
> Does `lxc image import server:fingerprint` work?

In https://github.com/canonical/lxd-ci/pull/240 I experimented with creating an image on the host and making it public. This would prove that image export over devlxd was working because the image would not be present anywhere else. The problem is that you can't specify the host as a remote (e.g. what do you put as "server" in the command you posted), so I tried out a few things related to [this comment](https://github.com/canonical/lxd/pull/13730#issuecomment-2219703206). The methods I tried were:
1. (naively) `lxc init "ubuntu:{host_image_fingerprint}" c1`. This doesn't work because the client tries to resolve the image against the ubuntu remote.
2. Using `lxc query -X POST /1.0/instances` and providing an image source with the fingerprint of the public image on the host. For this you need to provide a remote URL, so I added the ubuntu URL. This fails because the guest LXD tries to get the image from the remote to perform architecture checks ([here](https://github.com/canonical/lxd/blob/2445ed1a8fc945bba2dddd8bb1cbcc3f66bfae0b/lxd/instance/instance_utils.go#L617)).
3. Downloading the image directly via cURL. This doesn't really work because the downloaded `octet-stream` needs to be split into `rootfs` and metadata files before you can do anything. I also didn't like this because it doesn't actually test that the guest will download an image from the host when requested, only that the socket itself works.